File: src/main/java/voidchess/player/HumanPlayer.java
package voidchess.player;

import voidchess.board.ChessGameInterface;
import voidchess.helper.Move;
import voidchess.helper.PawnPromotion;
import voidchess.helper.Position;
import voidchess.ui.ChessGameUI;
import voidchess.ui.TableInterface;

import javax.swing.*;

/**
 * @author stephan
 */
public class HumanPlayer
        implements HumanPlayerInterface {
    private Position from;
    private ChessGameInterface game;
    private ChessGameUI ui;
    private TableInterface table;
    private boolean isWhitePlayer;
    private boolean isMyTurn;

    public HumanPlayer(TableInterface table,
                       boolean isWhitePlayer,
                       ChessGameUI ui,
                       ChessGameInterface game) {
        this.table = table;
        this.game = game;
        this.ui = ui;
        this.isWhitePlayer = isWhitePlayer;
        isMyTurn = false;
        ui.addPlayer(this);
    }

    public void play() {
        isMyTurn = true;
    }

    public void move(Move move) {
        dropMarkedPositions();
        isMyTurn = false;
        table.move(move);
    }

    public void mouseMovedOver(Position pos) {
        if (!isMyTurn) return;

        if (from == null) {
            if (game.isSelectable(pos, isWhitePlayer)) {
                ui.markPosition(pos, true);
            } else {
                ui.markPosition(null, true);
            }
        } else {
            if (game.isMoveable(from, pos, isWhitePlayer)) {
                ui.markPosition(pos, false);
            } else {
                ui.markPosition(null, false);
            }
        }
    }

    public void mouseClickedOn(Position pos) {
        if (!isMyTurn) return;

        if (from == null) {
            if (game.isSelectable(pos, isWhitePlayer)) {
                from = pos;
            }
        } else {
            if (game.isMoveable(from, pos, isWhitePlayer)) {
                move(Move.Companion.get(from, pos));
            }
        }
    }

    public PawnPromotion askForPawnPromotionType(Position pawnPosition) {
        String[] figs = {"Queen", "Rock", "Bishop", "Knight"};
        String out = (String) JOptionPane.showInputDialog(null,
                "Promote pawn to what type?",
                "pawn promotion",
                JOptionPane.QUESTION_MESSAGE,
                null,
                figs,
                "Queen");

        switch (out) {
            case "Queen":
                return PawnPromotion.QUEEN;
            case "Rock":
                return PawnPromotion.ROCK;
            case "Knight":
                return PawnPromotion.KNIGHT;
            default:
                return PawnPromotion.BISHOP;
        }
    }

    public void setIsPlaying(boolean isPlaying) {
        isMyTurn = isPlaying && isWhitePlayer;

        if (!isPlaying) {
            //removes selected fields even if game was aborted
            dropMarkedPositions();
        }
    }

    private void dropMarkedPositions() {
        from = null;
        ui.markPosition(null, true);
        ui.markPosition(null, false);
    }

    public void setColor(boolean isWhite) {
        isWhitePlayer = isWhite;
    }
}

File: src/main/kotlin/voidchess/player/HumanPlayer.kt
package voidchess.player

import voidchess.board.ChessGameInterface
import voidchess.board.MoveResult
import voidchess.helper.Move
import voidchess.helper.PawnPromotion
import voidchess.helper.Position
import voidchess.ui.ChessboardComponent
import voidchess.ui.PosType
import voidchess.ui.TableInterface

import javax.swing.*


class HumanPlayer(
        private val table: TableInterface,
        private var isWhitePlayer: Boolean,
        private val ui: ChessboardComponent,
        private val game: ChessGameInterface
) : HumanPlayerInterface {
    private var mouseHoverWhileInactivePos: Position? = null
    private var from: Position? = null
    private var isMyTurn: Boolean = false

    init {
        ui.setPlayer(this)
    }

    override fun play() {
        isMyTurn = true
        mouseMovedOver(mouseHoverWhileInactivePos)
    }

    fun move(move: Move) {
        dropMarkedPositions()
        isMyTurn = false
        table.move(move)
    }

    override fun mouseMovedOver(pos: Position?) {
        if (!isMyTurn) {
            mouseHoverWhileInactivePos = pos
            return
        }

        val lockedFrom = from
        if( pos==null) {
            ui.unmarkPosition(if(lockedFrom==null) PosType.HOVER_FROM else PosType.HOVER_TO)
            return
        }

        if (lockedFrom == null) {
            ui.unmarkPosition(PosType.HOVER_FROM)
            if (game.isSelectable(pos, isWhitePlayer)) {
                ui.markPosition(pos, PosType.HOVER_FROM)
            }
        } else {
            ui.unmarkPosition(PosType.HOVER_TO)
            if (game.isMoveable(lockedFrom, pos, isWhitePlayer)) {
                ui.markPosition(pos, PosType.HOVER_TO)
            }
        }
    }

    override fun mouseClickedOn(pos: Position) {
        if (!isMyTurn) return

        val lockedFrom = from
        if (lockedFrom == null) {
            if (game.isSelectable(pos, isWhitePlayer)) {
                from = pos
                ui.markPosition(pos, PosType.SELECT_FROM)
            }
        } else {
            if (game.isMoveable(lockedFrom, pos, isWhitePlayer)) {
                move(Move.get(lockedFrom, pos))
            }
        }
    }

    override fun askForPawnPromotionType(pawnPosition: Position): PawnPromotion {
        val figs = arrayOf("Queen", "Rook", "Bishop", "Knight")
        val out = JOptionPane.showInputDialog(null,
                "Promote pawn to what type?",
                "pawn promotion",
                JOptionPane.QUESTION_MESSAGE, null,
                figs,
                "Queen") as String

        return when (out) {
            "Queen" -> PawnPromotion.QUEEN
            "Rook" -> PawnPromotion.ROOK
            "Knight" -> PawnPromotion.KNIGHT
            else -> PawnPromotion.BISHOP
        }
    }

    override fun gameStarts() {
        isMyTurn = isWhitePlayer
    }

    override fun gameEnds(endoption: MoveResult, lastMoveByWhite: Boolean) {
        isMyTurn = false
        dropMarkedPositions()
    }

    private fun dropMarkedPositions() {
        from = null
        ui.unmarkPosition(PosType.HOVER_FROM)
        ui.unmarkPosition(PosType.SELECT_FROM)
        ui.unmarkPosition(PosType.HOVER_TO)
    }

    override fun setColor(isWhite: Boolean) {
        isWhitePlayer = isWhite
        ui.setViewPoint(isWhite)
    }
}
